### PR TITLE
lsp: Rename components, structs and enum from places where they are used

### PR DESCRIPTION
**Pull request by hunger**
This has pretty decent test coverage, but it is more involved than the first and it is admittedly a bit of a mess in the PR... 

I was test driven and forgot to stop along the way ;-)

### CODE REVIEW

**Comment by hunger** [comment]
It mostly works when testing from VSCode, but it still has a couple of rough edges where the tests pass, but VSCode does not like want to trigger. I'll poke at it a bit more tomorrow.

File: tools/lsp/common/rename_component.rs [hunk 1]
```diff
@@ -160,24 +160,40 @@ fn fix_imports(
...
 
+fn import_path(
+    document_directory: &Path,
+    import_specifier: &syntax_nodes::ImportSpecifier,
+) -> Option<PathBuf> {
+    let import = import_specifier
+        .child_token(SyntaxKind::StringLiteral)
+        .map(|t| t.text().trim_matches('"').to_string())?;
+
+    if import == "std-widgets.slint" || import == "std-widgets.s60" {
```
Comment (ogoffart) [inline tools/lsp/common/rename_component.rs:177]:
```suggestion
    if import == "std-widgets.slint" {
```
It was called `sixtyfps_widgets.60` but we no longer support that

File: tools/lsp/common/rename_component.rs [hunk 2]
```diff
@@ -1947,4 +2397,693 @@ export { Foo as User4Fxx }
...
+
+    if true: Rectangle {
+        Foo /* <- TEST_ME_8 */ { }
+    }
+
+    if false: Rectangle {
+        Foo /* <- TEST_ME_9 */ { }
+    }
+
+    function Foo(Foo: int) { Foo + 1; }
```
Comment (ogoffart) [inline tools/lsp/common/rename_component.rs:2466]: The test should also test that none of the `Foo` here matches with a component.

File: tools/lsp/common/rename_component.rs [hunk 3]
```diff
@@ -516,41 +544,495 @@ pub fn rename_identifier_from_declaration(
...
+
+#[derive(Clone, Debug)]
+struct DeclarationNodeQuery {
+    kind: DeclarationNodeQueryKind,
+    name: SmolStr,
+    node: SyntaxNode,
+}
+
+impl DeclarationNodeQuery {
+    fn new(node: SyntaxNode) -> Option<Self> {
```
Comment (ogoffart) [inline tools/lsp/common/rename_component.rs:852]: This seems like it is an incomplete duplication of `language::token_info::token_info()`
That function also do the lookup and tell you the exact type this is referring to.
So I think it would be a better candidate.

**Comment by hunger** [comment]
I started to pull out a bit of code from here into easier to review portions in separate PRs...

File: tools/lsp/common/rename_component.rs (hunk 1 above)
Comment (hunger) [reply tools/lsp/common/rename_component.rs:177]: I replaced this with

```rust
if import == "std-widgets.slint" || import.starts_with('@') {
```

We should not rename things imported from libraries in general.

File: tools/lsp/common/rename_component.rs (hunk 2 above)
Comment (hunger) [reply tools/lsp/common/rename_component.rs:2466]: None of this should trigger for function names yet, but true, that should be tested.

File: tools/lsp/common/rename_component.rs (hunk 3 above)
Comment (hunger) [reply tools/lsp/common/rename_component.rs]: Latest push contains the work to redo this using token_info.

I have an additional PR ready that adds renaming properties... but that is entirely untested so far and also needs to be integrated into the property editor. But this is extensible:-)

File: tools/lsp/common/token_info.rs [hunk 4]
```diff
@@ -186,6 +191,62 @@ pub fn token_info(document_cache: &common::DocumentCache, token: SyntaxToken) ->
...
+                    ElementType::Component(component)
+                        if component
+                            .node
+                            .as_ref()
+                            .map(|n| n.text_range().contains_range(token.text_range()))
+                            .unwrap_or_default() =>
+                    {
+                        return Some(TokenInfo::ElementType(element_type));
+                    }
+                    _ => { /* nothing to do */ }
```
Comment (ogoffart) [inline tools/lsp/common/token_info.rs:212]: Should that be a `if let`

File: tools/lsp/common/rename_component.rs [hunk 5]
```diff
@@ -329,310 +268,491 @@ fn fix_exports(
...
+                common::token_info::TokenInfo::NamedReference(s),
+                common::token_info::TokenInfo::NamedReference(o),
+            ) => s == o,
+            (
+                common::token_info::TokenInfo::EnumerationValue(s),
+                common::token_info::TokenInfo::EnumerationValue(o),
+            ) => s == o,
+            (
+                common::token_info::TokenInfo::LocalProperty(s),
+                common::token_info::TokenInfo::LocalProperty(o),
```
Comment (ogoffart) [inline tools/lsp/common/rename_component.rs:578]: I wonder if LocalProperty and LocalCallback can also be the same as a NamedReference or IncompleteNamedReference

**Review by ogoffart — COMMENTED** [review]
(partial review)

Nice to see this coming along.

Please make sure there are also tests that control that things that have the same name but are different (eg, a struct and a property) are not renamed. 
Apologies is this is already there and I missed it.

File: tools/lsp/common/rename_component.rs (hunk 5 above)
Comment (hunger) [reply tools/lsp/common/rename_component.rs:578]: I tired to come up with some valid comparison function and probably did some useless ones...

E.g. for the property renaming I played with yesterday evening I need to compare a `LocalProperty` with `NamedReference`. I'll just remove all the comparisons not yet used (and tested ;-).

File: tools/lsp/common/token_info.rs (hunk 4 above)
Comment (hunger) [reply tools/lsp/common/token_info.rs:212]: Yeap, that is probably nice, the next push will have that.

File: tools/lsp/common/rename_component.rs [hunk 6]
```diff
@@ -715,105 +835,266 @@ mod tests {
...
         assert_eq!(edited_text.len(), 1);
-        assert!(edited_text[0].contents.contains("component Foo /* <- TEST_ME_1 "));
         assert!(edited_text[0].contents.contains("export { XxxYyyZzz }"));
         assert!(edited_text[0].contents.contains("enum Xyz { Foo,"));
         assert!(edited_text[0].contents.contains("struct Abc { Foo:"));
-        assert!(edited_text[0].contents.contains("Foo /* Baz "));
+        assert!(edited_text[0].contents.contains("component XxxYyyZzz /* <- TEST_ME_1 "));
+        assert!(edited_text[0].contents.contains("TEST_ME_1 */ inherits Rectangle "));
+        assert!(edited_text[0].contents.contains("XxxYyyZzz /* <- TEST_ME_2 "));
+        assert!(edited_text[0].contents.contains("struct Foo {"));
```
Comment (hunger) [inline tools/lsp/common/rename_component.rs:1086]: This covers a struct not being renamed when a component of the same name is.

File: tools/lsp/common/rename_component.rs [hunk 7]
```diff
@@ -715,105 +835,266 @@ mod tests {
...
+            .contains("component Bar inherits XxxYyyZzz /* <- TEST_ME_3 "));
         assert!(edited_text[0].contents.contains("XxxYyyZzz /* <- TEST_ME_4 "));
-        assert!(edited_text[0].contents.contains("Foo := Baz {"));
         assert!(edited_text[0].contents.contains("XxxYyyZzz /* <- TEST_ME_5 "));
+        assert!(edited_text[0].contents.contains("Foo := Baz "));
         assert!(edited_text[0].contents.contains("XxxYyyZzz /* <- TEST_ME_6 "));
-        assert!(edited_text[0].contents.contains("function Foo(Foo: int) { Foo + 1; }"));
-        assert!(edited_text[0].contents.contains("function F() { self.Foo(42); }"));
         assert!(edited_text[0].contents.contains("XxxYyyZzz /* <- TEST_ME_7 "));
+        assert!(edited_text[0].contents.contains("function Foo(Foo:"));
```
Comment (hunger) [inline tools/lsp/common/rename_component.rs:1095]: ... function and parameters here...

**Review by ogoffart — APPROVED** [review]
Nice work!
Good to see this finally done.

**Comment by hunger** [comment]
I'll do one more round through CI: I seem to have missed a test for renaming a component with an enum of the same name in scope.
